scipy/linalg/_matfuncs_expm.h

```diff
@@ -0,0 +1,236 @@
+void pick_pade_structure_z(EXPM_Z* Am, const Py_ssize_t n, int* m, int* s);
+
+void pade_UV_calc_s(float* Am, const Py_ssize_t n, const int m, int* info);
+void pade_UV_calc_d(double* Am, const Py_ssize_t n, const int m, int* info);
+void pade_UV_calc_c(EXPM_C* Am, const Py_ssize_t n, const int m, int* info);
+void pade_UV_calc_z(EXPM_Z* Am, const Py_ssize_t n, const int m, int* info);
+
+
+#define PYERR(errobj,message) {PyErr_SetString(errobj,message); return NULL;}
+
+static PyObject* expm_error;
+
+static char doc_pps[] = "[m, s] = pick_pade_structure(Am)";
+
+static PyObject*
+pick_pade_structure(PyObject *dummy, PyObject *args) {
+    Py_ssize_t n;
+    int m = 0, s = 0;
+    PyArrayObject *ap_Am=NULL;
+
+    if (!PyArg_ParseTuple(args, ("O!"), &PyArray_Type, (PyObject **)&ap_Am)) return NULL;
+
+    if (!(PyArray_IS_C_CONTIGUOUS(ap_Am)) ||
+        ((PyArray_TYPE(ap_Am) != NPY_FLOAT64) &&
+         (PyArray_TYPE(ap_Am) != NPY_FLOAT32) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX64) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX128)) ||
+        (PyArray_NDIM(ap_Am) != 3)
+        )
+    {
+        PYERR(expm_error, "Am must be a 3D C-contiguous array with size (5, n, n) "
+                          " that of type float32, float64, complex64, or complex128.");
+    }
+
+    n = PyArray_DIMS(ap_Am)[2];
+    // Make the call based on dtype
+    switch (PyArray_TYPE(ap_Am))
+    {
+        case (NPY_FLOAT32):
+        {
+            float* Ams = (float*)PyArray_DATA(ap_Am);
+            pick_pade_structure_s(Ams, n, &m, &s);
+            break;
+        }
+        case (NPY_FLOAT64):
+        {
+            double* Amd = (double*)PyArray_DATA(ap_Am);
+            pick_pade_structure_d(Amd, n, &m, &s);
+            break;
+        }
+        case (NPY_COMPLEX64):
+        {
+            EXPM_C* Amc = (EXPM_C*)PyArray_DATA(ap_Am);
+            pick_pade_structure_c(Amc, n, &m, &s);
+            break;
+        }
+        case (NPY_COMPLEX128):
+        {
+            EXPM_Z* Amz = (EXPM_Z*)PyArray_DATA(ap_Am);
+            pick_pade_structure_z(Amz, n, &m, &s);
+            break;
+        }
+    }
+
+    return Py_BuildValue("ii", m, s);
+}
+
+
+static char doc_puv[] = "info = pade_UV_calc(Am, m)";
+
+static PyObject*
+pade_UV_calc(PyObject *dummy, PyObject *args) {
+    int m = 0, info = 0;
+    Py_ssize_t n;
+    PyArrayObject *ap_Am=NULL;
+
+    if (!PyArg_ParseTuple(args, ("O!i"), &PyArray_Type, (PyObject **)&ap_Am, &m)) return NULL;
+    if (!(PyArray_IS_C_CONTIGUOUS(ap_Am)) ||
+        ((PyArray_TYPE(ap_Am) != NPY_FLOAT64) &&
+         (PyArray_TYPE(ap_Am) != NPY_FLOAT32) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX64) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX128)) ||
+        (PyArray_NDIM(ap_Am) != 3)
+        )
+    {
+        PYERR(expm_error, "Am must be a 3D C-contiguous array with size (5, n, n) "
+                          " that of type float32, float64, complex64, or complex128.");
+    }
```

Comment:
This is the same as in `pick_pade_structure` and could be factored out.

Reply:
Indeed, I just didn't want to deal with propagating the error from a nested function.